src/OrbitBase/SimpleExecutorTest.cpp

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <gtest/gtest.h>
+
+#include "OrbitBase/SimpleExecutor.h"
+
+namespace orbit_base {
+
+TEST(SimpleExecutor, ScheduledTaskShouldBeCalledSimpleWithVoid) {
+  auto executor = SimpleExecutor::Create();
+  bool called = false;
+  auto future = executor->Schedule([&called]() { called = true; });
+  executor->ExecuteScheduledTasks();
+  EXPECT_TRUE(called);
+  EXPECT_TRUE(future.IsFinished());
+}
+
+TEST(SimpleExecutor, ScheduledTaskShouldBeCalledSimpleWithInt) {
+  auto executor = SimpleExecutor::Create();
+  bool called = false;
+  auto future = executor->Schedule([&called]() {
+    called = true;
+    return 42;
+  });
+  executor->ExecuteScheduledTasks();
+  EXPECT_TRUE(called);
+  ASSERT_TRUE(future.IsFinished());
+  EXPECT_EQ(future.Get(), 42);
+}
+
+TEST(SimpleExecutor, ChainedTaskedShouldBeCalledSimple) {
+  auto executor = SimpleExecutor::Create();
+  bool called = false;
+  Promise<void> promise{};
+  auto future = promise.GetFuture();
+  auto chained_future = executor->ScheduleAfter(future, [&called]() { called = true; });
+  EXPECT_FALSE(called);
+  EXPECT_FALSE(chained_future.IsFinished());
+  promise.MarkFinished();
+  executor->ExecuteScheduledTasks();
+  EXPECT_TRUE(called);
+  EXPECT_TRUE(chained_future.IsFinished());
+}
+
+TEST(SimpleExecutor, ScheduleAfterIfSuccessShortCircuitOnErrorVoid) {
+  auto executor = SimpleExecutor::Create();
+  bool called = false;
+  Promise<ErrorMessageOr<void>> promise{};
+  auto future = promise.GetFuture();
+  auto chained_future = executor->ScheduleAfterIfSuccess(future, [&called]() { called = true; });
+  EXPECT_FALSE(called);
+  EXPECT_FALSE(chained_future.IsFinished());
+
+  constexpr const char* const kErrorMessage{"Error"};
+  promise.SetResult(ErrorMessage{kErrorMessage});
+  executor->ExecuteScheduledTasks();
+  EXPECT_FALSE(called);
+  ASSERT_TRUE(chained_future.IsFinished());
+  EXPECT_TRUE(chained_future.Get().has_error());
+  EXPECT_EQ(chained_future.Get().error().message(), kErrorMessage);
+}
+
+TEST(SimpleExecutor, ScheduleAfterIfSuccessShortCircuitOnErrorInt) {
+  auto executor = SimpleExecutor::Create();
+  bool called = false;
+  Promise<ErrorMessageOr<int>> promise{};
+  auto future = promise.GetFuture();
+  auto chained_future = executor->ScheduleAfterIfSuccess(future, [&called](int value) {
+    EXPECT_EQ(value, 42);
+    called = true;
+    return 1 + value;
+  });
+  executor->ExecuteScheduledTasks();
```

Comment:
At this point the lambda is still in `Executor::waiting_continuations`, correct? I wonder if checking the value of `GetNumberOfWaitingContinuations` here, is a good or a bad idea. I think it might help explaining whats going on to the reader of this test, but it also unnecessarily ties this test suite to the implementation of the `Executor` base class.

Reply:
There is also another point to consider: Ideally we wanna have orthogonal tests, meaning whenever I introduce a single bug, ideally I want to have a single test fail. This will allow me to pinpoint the problem very quickly. Of course that's not always possible, or let's say most of the time that's very hard because of dependencies.

But whenever possible I prefer not to test behaviour that is already covered by another test. In this case it's covered by the `Executor.*` tests.

Needless to say that these tests are far away from consistently following this guideline because most of them are just copied from what has already been there. But I tried at least not to introduce more double-coverage.

Nevertheless, if you believe this test (these tests) requires some explanation for the reader I can totally improve on that. I guess in that case comments would be the way to go. WDYT?

Reply:
Using the term orthogonal, plus the explanation is a great way to put in words what was a somewhat fuzzy feeling in my head. I have to remember that. 

So yes, I agree, lets leave the tests as is. I just tried to come up with explaining comments, but at this point I think I fully understood all the code and then the tests also are clear the way they currently are. So im going to resolve the conversation